cupyx/scipy/sparse/_index.py

```diff
@@ -0,0 +1,388 @@
+        if isinstance(row, _int_scalar_types):
+            if isinstance(col, _int_scalar_types):
+                return self._get_intXint(row, col)
+            elif isinstance(col, slice):
+                return self._get_intXslice(row, col)
+            elif col.ndim == 1:
+                return self._get_intXarray(row, col)
+            raise IndexError('index results in >2 dimensions')
+        elif isinstance(row, slice):
+            if isinstance(col, _int_scalar_types):
+                return self._get_sliceXint(row, col)
+            elif isinstance(col, slice):
+                if row == slice(None) and row == col:
+                    return self.copy()
+                return self._get_sliceXslice(row, col)
+            elif col.ndim == 1:
+                return self._get_sliceXarray(row, col)
+            raise IndexError('index results in >2 dimensions')
+        elif row.ndim == 1:
+            if isinstance(col, _int_scalar_types):
+                return self._get_arrayXint(row, col)
+            elif isinstance(col, slice):
+                return self._get_arrayXslice(row, col)
+        else:  # row.ndim == 2
+            if isinstance(col, _int_scalar_types):
+                return self._get_arrayXint(row, col)
+            elif isinstance(col, slice):
+                raise IndexError('index results in >2 dimensions')
+            elif row.shape[1] == 1 and (col.ndim == 1 or col.shape[0] == 1):
+                # special case for outer indexing
+                return self._get_columnXarray(row[:, 0], col.ravel())
+
+        # The only remaining case is inner (fancy) indexing
+        row, col = cupy.broadcast_arrays(row, col)
+        if row.shape != col.shape:
+            raise IndexError('number of row and column indices differ')
+        if row.size == 0:
+            return self.__class__(cupy.atleast_2d(row).shape, dtype=self.dtype)
+        return self._get_arrayXarray(row, col)
+
+    def _parse_indices(self, key):
+        M, N = self.shape
+        row, col = _unpack_index(key)
+
+        # Scipy calls sputils.isintlike() rather than
+        # isinstance(x, _int_scalar_types). Comparing directly to int
+        # here to minimize the impact of nested exception catching
+
+        if isinstance(row, _int_scalar_types):
+            row = _normalize_index(row, M, 'row')
+        elif not isinstance(row, slice):
+            row = self._asindices(row, M)
+
+        if isinstance(col, _int_scalar_types):
+            col = _normalize_index(col, N, 'column')
+        elif not isinstance(col, slice):
+            col = self._asindices(col, N)
+
+        return row, col
+
+    def _asindices(self, idx, length):
+        """Convert `idx` to a valid index for an axis with a given length.
+        Subclasses that need special validation can override this method.
+        """
+        try:
+            x = cupy.asarray(idx, dtype=self.indices.dtype)
+        except (ValueError, TypeError, MemoryError):
```

Comment:
Should an `IndexError` be reraised for `MemoryError`?

Reply:
This came from Scipy. My thinking is that this would be in the case where a particular row/col pattern has been repeated to the point where returning the entries at those indices would blow the memory, thus it's better to return an `IndexError` since the memory error resulted initially from the user's choice of indices.